### PR TITLE
IngressBackend UpstreamTrafficSetting validations

### LINKED ISSUE #4638: Add stricter validations for IngressBackend and UpstreamTrafficSetting resources
**Please describe the Improvement and/or Feature Request**
Add validations in the validating webhook to ensure duplicate resources targeting the same destination is not permitted, as these would result in policy conflicts.

- For IngressBackend, ensure multiple IngressBackends cannot specify the same backend service
- For UpstreamTrafficSetting, ensure multiple UpstreamTrafficSettings cannot specify the same upstream host.

**Scope (please mark with X where applicable)**
- Validations          [X]

**Possible use cases**
To prevent user misconfiguration.

### PR DESCRIPTION
**Pull request by keithmattix**
**Description**:
Fixes #4638. Adds stricter validatingWebhook validations for IngressBackend and UpstreamTrafficSetting (duplicate backends and hosts respectively). This PR also changes MeshService -> IngressBackend retrieval process to match on port as well as service name. This appears to have been an accidental omission from the original design since one would expect to be able to create an IngressBackend that allows access to the same service on different ports.

**Testing done**:
Additional unit tests added in the appropriate Golang packages

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| Control Plane              | [X] |
| Other                      | [X] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? no
    -   Did you notify the maintainers and provide attribution? no

2. Is this a breaking change? no

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)? N/A

### CODE REVIEW

File: pkg/validator/types.go [hunk 1]
```diff
@@ -2,6 +2,10 @@ package validator
 
 import (
 	"github.com/openservicemesh/osm/pkg/logger"
+	"github.com/pkg/errors"
 )
 
 var log = logger.New(ValidatorWebhookSvc)
+
+var ErrIngressBackendDuplicateBackends = errors.New("error: duplicate backends detected")
+var ErrUpstreamTrafficSettingDuplicateHosts = errors.New("error: duplicate hosts detected")
```
Comment (jaellio) [inline pkg/validator/types.go:11]:
```suggestion
var (
  ErrIngressBackendDuplicateBackends      = errors.New("error: duplicate backends detected")
  ErrUpstreamTrafficSettingDuplicateHosts = errors.New("error: duplicate hosts detected")
)
```

Comment (jaellio) [reply pkg/validator/types.go]: In most errors we create we don't use the format "error: \<err msg\>". Instead, we do something like the following `var ErrCouldNotOpen = errors.New("could not open")`

File: pkg/policy/client.go [hunk 2]
```diff
@@ -162,7 +162,8 @@ func (c client) GetIngressBackendPolicy(svc service.MeshService) *policyV1alpha1
 		// Multiple IngressBackend policies for the same backend will be prevented
 		// using a validating webhook.
 		for _, backend := range ingressBackend.Spec.Backends {
-			if backend.Name == svc.Name {
+			// we need to check ports to allow ingress to multiple ports on the same svc
+			if backend.Name == svc.Name && backend.Port.Number == int(svc.Port) && backend.Port.Protocol == svc.Protocol {
```
Comment (shashankram) [inline pkg/policy/client.go:166]: The TargetPort check is performed in catalog/ingress.go, but makes sense to be done here. I can't recollect why I did not add it here.
```suggestion
			if backend.Name == svc.Name && backend.Port.Number == int(svc.TargetPort) && backend.Port.Protocol == svc.Protocol {
```

File: pkg/validator/types.go (hunk 1 above)
Comment (keithmattix) [reply pkg/validator/types.go]: Ah, thanks for the heads up! Fixed

File: pkg/validator/server.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ import (
 	"github.com/openservicemesh/osm/pkg/certificate"
 	"github.com/openservicemesh/osm/pkg/constants"
 	"github.com/openservicemesh/osm/pkg/errcode"
+	policyClientset "github.com/openservicemesh/osm/pkg/gen/client/policy/clientset/versioned"
```
Comment (shashankram) [inline pkg/validator/server.go:20]: Group this with the API imports and leave a line after it.

```
stdlib

3rd party

OSM APIs

local pkg
```

File: pkg/validator/types.go [hunk 4]
```diff
@@ -2,6 +2,10 @@ package validator
 
 import (
 	"github.com/openservicemesh/osm/pkg/logger"
+	"github.com/pkg/errors"
 )
 
 var log = logger.New(ValidatorWebhookSvc)
+
+var errIngressBackendDuplicateBackends = errors.New("duplicate backends detected")
+var errUpstreamTrafficSettingDuplicateHosts = errors.New("duplicate hosts detected")
```
Comment (shashankram) [inline pkg/validator/types.go:11]: Let's directly return the errors instead of statically defining this so we can add more context to the error message.

File: pkg/validator/validators.go [hunk 5]
```diff
@@ -76,13 +86,25 @@ func trafficTargetValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Adm
...
 	if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(ingressBackend); err != nil {
 		return nil, err
 	}
 
+	type backendCacheKey struct {
+		name string
+		port int
+	}
+
+	backends := make(map[backendCacheKey]struct{}, len(ingressBackend.Spec.Backends))
```
Comment (shashankram) [inline pkg/validator/validators.go:100]: Use `mapset` here. If `mapset.Add()` returns true, it implies the item was added and does not contain a duplicate.

File: pkg/validator/validators.go [hunk 6]
```diff
@@ -136,9 +160,52 @@ func ingressBackendValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Ad
...
 	}
 
+	err := kc.checkDuplicateBackends(ingressBackend)
+	if errors.Is(err, errIngressBackendDuplicateBackends) {
+		return nil, err
+	}
+
+	// Fail open for transient errors, but log
+	if err != nil {
+		log.Error().Err(err).Msg("transient error encountered checking duplicate IngressBackend backends")
```
Comment (shashankram) [inline pkg/validator/validators.go:170]: This log message seems ambiguous. Could you reword this to remove the usage of the word `transient`.

File: pkg/validator/validators.go [hunk 7]
```diff
@@ -136,9 +160,52 @@ func ingressBackendValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Ad
...
+		}
+	}
+
+	if conflictString.Len() != 0 {
+		return fmt.Errorf("%w\n%s", errIngressBackendDuplicateBackends, conflictString.String())
+	}
+
+	return nil
+}
+
```
Comment (shashankram) [inline pkg/validator/validators.go:208]: The logic here can be simplified by checking for duplicates inline while processing the backend. Can something like this be done:
When you see a backend, check the cache to see if the backend is already referenced in some other ingress resource. If it is, return an error. Otherwise, the backend is new.

Why do we need to match the entire IngressBackend object with all other IngressBackend objects?

**Review by shashankram — CHANGES_REQUESTED** [review]
Thanks for adding these. Left a few comments, otherwise looks good.

File: pkg/validator/types.go (hunk 4 above)
Comment (keithmattix) [reply pkg/validator/types.go:11]: These error vars are wrapped [later on](https://github.com/openservicemesh/osm/pull/4640/files#diff-ca09fc1c4167db01ac42e4c120198cbd797d234b280b9986266ac0440b016289R203) so as to add more context. The main reason for explicitly creating types here is so that we can use `errors.Is()` in order to [decide](https://github.com/openservicemesh/osm/pull/4640/files#diff-ca09fc1c4167db01ac42e4c120198cbd797d234b280b9986266ac0440b016289R164) whether or not  to fail the validating webhook

File: pkg/validator/validators.go (hunk 6 above)
Comment (shashankram) [reply pkg/validator/validators.go:168]: The error path here should be deterministic. I see why it's not the case now as you are using `List()` below, but we could simply rely on the cache of objects already present in `policy/client.go`.

File: pkg/validator/validators.go [hunk 8]
```diff
@@ -2,18 +2,24 @@ package validator
...
+	policyClientset "github.com/openservicemesh/osm/pkg/gen/client/policy/clientset/versioned"
+	"github.com/openservicemesh/osm/pkg/policy"
 	"github.com/pkg/errors"
 	smiAccess "github.com/servicemeshinterface/smi-sdk-go/pkg/apis/access/v1alpha3"
 	smiSpecs "github.com/servicemeshinterface/smi-sdk-go/pkg/apis/specs/v1alpha4"
 	admissionv1 "k8s.io/api/admission/v1"
 
 	configv1alpha2 "github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
 	policyv1alpha1 "github.com/openservicemesh/osm/pkg/apis/policy/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (jaellio) [inline pkg/validator/validators.go:22]:
```suggestion
	admissionv1 "k8s.io/api/admission/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	configv1alpha2 "github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
	policyv1alpha1 "github.com/openservicemesh/osm/pkg/apis/policy/v1alpha1"
```

Comment (keithmattix) [reply pkg/validator/validators.go]: Fixed!

File: pkg/validator/validators.go (hunk 7 above)
Comment (keithmattix) [reply pkg/validator/validators.go]: Fixed as we discussed offline

File: pkg/validator/types.go (hunk 4 above)
Comment (keithmattix) [reply pkg/validator/types.go]: Removed since validators were made to be deterministic

File: pkg/validator/validators.go [hunk 9]
```diff
@@ -76,13 +86,24 @@ func trafficTargetValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Adm
...
+		port int
+	}
+
+	backends := mapset.NewSet()
 	for _, backend := range ingressBackend.Spec.Backends {
+		unique := backends.Add(setEntry{backend.Name, backend.Port.Number})
+		if !unique {
+			return nil, errors.Errorf("Duplicate backends detected with service name: %s and port: %d", backend.Name, backend.Port.Number)
+		}
+
```
[the remark this answers is not in the document]
Comment (keithmattix) [reply pkg/validator/validators.go:106]: Good catch - added to ingress backend but not upstreamtrafficsetting for readability

File: pkg/validator/validators.go [hunk 10]
```diff
@@ -136,9 +156,53 @@ func ingressBackendValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Ad
...
+	if err != nil {
+		return nil, err
+	}
+
 	return nil, nil
 }
 
+// checkDuplicateBackends checks to see if the backends present in the given
+// ingressBackend conflict with existing ingressBackends
+func (kc *policyValidator) checkDuplicateBackends(candidateIngressBackend *policyv1alpha1.IngressBackend) error {
```
Comment (shashankram) [inline pkg/validator/validators.go:169]: We shouldn't need this if we use the API in policy.Controller to lookup an existing IngressBackend matching the same backend?

File: pkg/validator/validators.go [hunk 11]
```diff
@@ -182,6 +246,54 @@ func egressValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionR
...
+	}
+
+	err := kc.checkDuplicateHosts(upstreamTrafficSetting)
+	if errors.Is(err, errUpstreamTrafficSettingDuplicateHosts) {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (shashankram) [inline pkg/validator/validators.go:268]: Let's replace this with a check `GetUpstreamTrafficSetting(opt)` where opt is of type UpstreamTrafficSettingGetOpt and supports lookups in the case based on host. This means you need to add a new field `Host` to the `UpstreamTrafficSettingGetOpt` type.

Once done, this can be simplified to:
```
if obj := policyController.GetUpstreamTrafficSetting(opt); obj != nil {
    // duplicate
}
```

File: pkg/validator/validators.go (hunk 10 above)
Comment (keithmattix) [reply pkg/validator/validators.go]: Ah yep, I was duplicating the loop; I see now where `ingressBackendValidator` already loops through the list of backends. The end result is much simpler 👍🏾

File: pkg/validator/validators.go [hunk 12]
```diff
@@ -76,13 +86,46 @@ func trafficTargetValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Adm
...
+		}
+
+		fakeMeshSvc := service.MeshService{
+			Name:       backend.Name,
+			TargetPort: uint16(backend.Port.Number),
+			Protocol:   backend.Port.Protocol,
+		}
+
+		if matchingPolicy := kc.policyClient.GetIngressBackendPolicy(fakeMeshSvc); matchingPolicy != nil {
+			// we've found a duplicate
```
Comment (shashankram) [inline pkg/validator/validators.go:116]: Another check is required here to ensure `matchingPolicy.Name != ingressBackend.Name`, which would be the case for an `Update` API validation.

**Review by shashankram — COMMENTED** [review]
Looks great, just 1 major comment to address.

File: pkg/validator/validators.go (hunk 12 above)
Comment (keithmattix) [reply pkg/validator/validators.go:116]: Ah, fantastic catch - I'd forgotten that validation webhooks are run on update as well. I fixed these and added test cases to confirm that they work as expected 👍🏾

File: pkg/policy/client.go [hunk 13]
```diff
@@ -162,7 +162,8 @@ func (c client) GetIngressBackendPolicy(svc service.MeshService) *policyV1alpha1
 		// Multiple IngressBackend policies for the same backend will be prevented
 		// using a validating webhook.
 		for _, backend := range ingressBackend.Spec.Backends {
-			if backend.Name == svc.Name {
+			// we need to check ports to allow ingress to multiple ports on the same svc
+			if backend.Name == svc.Name && backend.Port.Number == int(svc.TargetPort) && backend.Port.Protocol == svc.Protocol {
```
Comment (shashankram) [inline pkg/policy/client.go:166]: Let's drop the protocol check here as it breaks existing behavior. The protocol on a service represents its protocol for east-west traffic (pod to pod), not north-south (external to pod). So a service with `http` as its internal mesh protocol can still use `https` as the protocol with ingress.

File: pkg/policy/client.go [hunk 14]
```diff
@@ -208,6 +209,10 @@ func (c client) GetUpstreamTrafficSetting(options UpstreamTrafficSettingGetOpt)
 	for _, resource := range c.caches.upstreamTrafficSetting.List() {
 		upstreamTrafficSetting := resource.(*policyV1alpha1.UpstreamTrafficSetting)
 
+		if upstreamTrafficSetting.Spec.Host == options.Host {
```
Comment (shashankram) [inline pkg/policy/client.go:212]: Need a corresponding check in L194.

File: pkg/validator/validators.go [hunk 15]
```diff
@@ -76,13 +86,46 @@ func trafficTargetValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Adm
...
+		port int
+	}
 
+	backends := mapset.NewSet()
+	var conflictString strings.Builder
+	conflictingIngressBackends := mapset.NewSet()
 	for _, backend := range ingressBackend.Spec.Backends {
+		if unique := backends.Add(setEntry{backend.Name, backend.Port.Number}); !unique {
+			return nil, errors.Errorf("Duplicate backends detected with service name: %s and port: %d", backend.Name, backend.Port.Number)
+		}
```
Comment (shashankram) [inline pkg/validator/validators.go:107]: Why do we need this? Can't this be simplified to returning early in case of a duplicate:
```
for ingressBackend in ingressBackendList:
  if duplicate detected:
      return error(...)
```

**Review by shashankram — CHANGES_REQUESTED** [review]

File: pkg/validator/validators.go [hunk 16]
```diff
@@ -108,6 +151,10 @@ func ingressBackendValidator(req *admissionv1.AdmissionRequest) (*admissionv1.Ad
 		}
 	}
 
+	if conflictString.Len() != 0 {
+		return nil, fmt.Errorf("duplicate backends detected\n%s", conflictString.String())
+	}
+
```
[the remark this answers is not in the document]
Comment (keithmattix) [reply pkg/validator/validators.go:157]: @shashankram This and your previous comment seem to be talking about the same thing, so I'm just going to respond to it once. My thought process was that a more detailed error message that tells the customer all of their duplicates at once is a better developer experience than having to apply, fix, apply, fix, etc. I don't personally feel like the code is too complicated, but I'm naturally biased. I'm not too tied to the implementation either way, so if you feel strongly about it, I can change it.

[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/validator/validators.go:157]: Not a blocker. Feel free to ignore.

**Review by shashankram — APPROVED** [review]

**Review by jaellio — APPROVED** [review]
